### PR TITLE
refactor lsp_buf_request_sync to also have an async method

### PR DESCRIPTION
**Pull request by oberblastmeister**
`lsp_buf_request_sync` needs an async method that does the same thing. However, `buf_request` is not an alternative to `lsp_buf_request_sync` because it accepts a callback that will be called multiple times for each client that has a response. I have added a new method called `lsp_buf_request_all` that is the same as `lsp_buf_request_sync` but is async. This way, `lsp_buf_request_sync` can be implemented in terms of the async method using `vim.wait`.

### CODE REVIEW

File: runtime/lua/vim/lsp.lua [hunk 1]
```diff
@@ -1255,26 +1299,21 @@ end
...
+
+  local cancel = lsp.buf_request_all(bufnr, method, params, function(it)
+    request_results = it
+  end)
 
   local wait_result, reason = vim.wait(timeout_ms or 100, function()
-    return result_count >= expected_result_count
-  end, 10)
+    return request_results ~= nil
+  end)
```
Comment (mjlbach) [inline runtime/lua/vim/lsp.lua:1310]: I believe the implicit default for this will now be 200 ms (over the default 100 ms timeout window) is that what you wanted?

Comment (oberblastmeister) [reply runtime/lua/vim/lsp.lua:1310]: Thats a mistake

File: runtime/lua/vim/lsp.lua [hunk 2]
```diff
@@ -273,8 +273,12 @@ end
 --@returns (function) Memoized function
 local function once(fn)
```
Comment (mjlbach) [inline runtime/lua/vim/lsp.lua:274]: Why is this change necessary?

File: runtime/lua/vim/lsp.lua [hunk 3]
```diff
@@ -1239,9 +1243,49 @@ function lsp.buf_request(bufnr, method, params, handler)
   return client_request_ids, _cancel_all_requests
 end
 
---- Sends a request to a server and waits for the response.
+---Sends an async request for all active clients attached to the buffer.
```
[the remark this answers is not in the document]
Comment (oberblastmeister) [reply runtime/lua/vim/lsp.lua:1246]: > this documentation needs to be improved a bit. Technically all requests to the language server are "async"

I used the same wording as `buf_request`. Obviously all requests are 'async' but if we use `vim.wait` then it is not, so I think it is okay to make that distinction.

> This is a little wordy, I would say something like `Executes the callback on the pooled result

That is better wording, will change it

File: runtime/lua/vim/lsp.lua (hunk 2 above)
Comment (oberblastmeister) [reply runtime/lua/vim/lsp.lua:274]: If a function is just used for side-effects `once` will not run it once

Comment (mjlbach) [reply runtime/lua/vim/lsp.lua:274]: Ah, I guess that's fine then.

File: runtime/lua/vim/lsp.lua [hunk 4]
```diff
@@ -1239,9 +1243,49 @@ function lsp.buf_request(bufnr, method, params, handler)
...
+    request_results[client_id] = { error = err, result = result }
+    result_count = result_count + 1
+    set_expected_result_count()
+
+    if result_count >= expected_result_count then
+      callback(request_results)
+    end
+  end
+
+  client_request_ids, cancel = lsp.buf_request(bufnr, method, params, _sync_handler)
```
Comment (mjlbach) [inline runtime/lua/vim/lsp.lua:1281]: Is it always guaranteed that a function which accepts a callback in luv returns before the callback has executed? otherwise client_request_ids is nil when set_expected result_count is called by the _sync_handler callback.

Comment (oberblastmeister) [reply runtime/lua/vim/lsp.lua:1283]: Yes I think it should always but I'm not sure

Comment (mjlbach) [reply runtime/lua/vim/lsp.lua:1283]: If you can confirm that would be great :) I can look into it later too. Otherwise, the PR looks good to me (thanks!). 

I'm guessing you have some goal/utility in mind for lsp.buf_request_all as we don't really use buf_request_sync all that much, and this PR doesn't really affect it. So I'm fine with merging after we confirm no unexpected nils.

File: runtime/lua/vim/lsp.lua [hunk 5]
```diff
@@ -267,14 +267,20 @@ end
 
 --@private
 --- Memoizes a function. On first run, the function return value is saved and
---- immediately returned on subsequent runs.
+--- immediately returned on subsequent runs. If the function returns a multival,
+--- only the first one will be memoized and returned. The function will only be run once
```
Comment (mjlbach) [inline runtime/lua/vim/lsp.lua:271]: maybe returned value instead of one

Comment (mjlbach) [reply runtime/lua/vim/lsp.lua:271]: And should be a comma after once

Comment (oberblastmeister) [reply runtime/lua/vim/lsp.lua]: > maybe returned value instead of one

?

Comment (mjlbach) [reply runtime/lua/vim/lsp.lua]: If the function returns a multival,only the first {one -> returned value}

File: runtime/lua/vim/lsp.lua (hunk 4 above)
Comment (oberblastmeister) [reply runtime/lua/vim/lsp.lua:1283]: We will use `buf_request_all` in telescope because `buf_request_sync` is synchronous

Comment (mjlbach) [reply runtime/lua/vim/lsp.lua:1283]: @tjdevries Do you know re: callback before return result?

**Comment by oberblastmeister** [comment]
Can we make `once` not private? I think it is a useful function that would be good as `vim.once`

**Comment by mjlbach** [comment]
Sure, but I would file a separate PR.

**Comment by mjlbach** [comment]
nvim-salsa when?!

**Comment by oberblastmeister** [comment]
Nice one

File: runtime/lua/vim/lsp.lua (hunk 4 above)
Comment (oberblastmeister) [reply runtime/lua/vim/lsp.lua:1283]: I asked tj on stream, he says yes but he never proved it

**Comment by mjlbach** [comment]
Ok, well I blame TJ if we get a bunch of random `nil`-sayers filing issues. Squash your commits (I'd recommend a commit message like `lsp: add lsp.buf_request_all for invoking asynchronous callbacks`

**Comment by oberblastmeister** [comment]
oh no not this again. Im so bad at git :smile:

**Comment by mjlbach** [comment]
Lol just 'git rebase -i master' and change all but the top commit from pick to squash, and change the top commit to reword.

**Comment by oberblastmeister** [comment]
Okay I think I rebased correctly

**Comment by mjlbach** [comment]
Yep! Gj, I will merge when tests are green.
